rust/otap-dataflow/crates/engine/src/exporter.rs

```diff
@@ -87,106 +266,175 @@ impl<PData> MessageChannel<PData> {
         MessageChannel {
             control_rx,
             pdata_rx,
+            shutting_down_deadline: None,
+            pending_shutdown: None,
+            shutting_down: false,
         }
     }
 
     /// Asynchronously receives the next message to process.
-    /// Control messages are prioritized over pdata messages.
+    ///
+    /// Order of precedence:
+    ///
+    /// 1. Before a `Shutdown` is seen: control messages are always
+    ///    returned ahead of pdata.
+    /// 2. After the first `Shutdown` is received:
+    ///    - All further control messages are silently discarded.
+    ///    - Pending pdata are drained until the shutdown deadline.
+    /// 3. When the deadline expires (or was `0`): the stored `Shutdown` is returned.
+    ///    Subsequent calls return `RecvError::Closed`.
     ///
     /// # Errors
     ///
-    /// Returns a [`RecvError`] if both channels are closed.
+    /// Returns a [`RecvError`] if both channels are closed, or if the
+    /// shutdown deadline has passed.
     pub async fn recv(&mut self) -> Result<Message<PData>, RecvError> {
-        tokio::select! {
-            biased;     // Instruct the select macro to poll the futures in the order they appear
-                        // from top to bottom.
+        let mut sleep_until_deadline: Option<Pin<Box<Sleep>>> = None;
 
-            // Prioritize control messages explicitly
-            control_res = self.control_rx.recv() => {
-                control_res.map(|ctrl| Message::Control(ctrl))
+        loop {
+            if self.shutting_down {
+                return Err(RecvError::Closed);
```

Comment:
It would be nice to close the pdata and control data receivers here to notify the respective senders if they still try to send the messages.

Reply:
Good idea. Done.